File: src/mitgcm_inputs/Kext_for_GoT_iNEST_V2.py
# from legacy code Kext_for_GoT_iNEST_V2.m
import matplotlib.pyplot as plt
import numpy as np
from scipy.interpolate import griddata

import mitgcm_inputs.pathes_to_data as ptd
from mitgcm_inputs.parameters import is_plot


def plot2D(is_plot, A):
    if is_plot:
        plt.figure()
        plt.pcolor(A)
        plt.colorbar()
        plt.show()
    else:
        print("Not plotting figure")
    return 0


print("Start Kext_for_GoT_iNEST_V2")

if not is_plot:
    print("No figures printed")
# Path to files
data_file = ptd.med_data_path
strdir_grid = ptd.med_grid_path
mask_file_got = ptd.bathymetry_path + "/hFacC.data"

# parameters mediterran sea
nx = 570
ny = 264
nz = 75
days = 1921

# parameters Gulf of Trieste
yc_origin = 45.4
xc_origin = 13.22
delX = 1 / 768
delY = 1 / 768
nx_got = 450
ny_got = 300
nz_got = 40
# load mask for GoT
fr2 = open(mask_file_got, "rb")
mask_got = np.fromfile(fr2, dtype=np.float32).reshape((nz_got, ny_got, nx_got))
fr2.close()
mask_got = mask_got[0, :, :]
print("SHAPE GoT", np.shape(mask_got))
plot2D(is_plot, mask_got)

fhfacC = open(f"{strdir_grid}Computational_hFacC", "rb")
mask_hfac = np.fromfile(fhfacC, dtype=">f4").reshape((nz, ny, nx))
fhfacC.close()

mask_hfac[mask_hfac != 1] = 0
mask_hfac = mask_hfac[0, :, :]
print("SHAPE Med", np.shape(mask_hfac))

plot2D(is_plot, mask_hfac)
kext = np.zeros((days, ny, nx))
fKext = open(data_file, "rb")
A = np.zeros((days, ny, nx))
A = np.fromfile(fKext, dtype=">f4", count=days * nx * ny).reshape(
    (days, ny, nx)
)
fKext.close()
for idays in range(days):
    kext[idays, :, :] = A[0, :, :] * mask_hfac
plot2D(is_plot, kext[0, :, :])

mask_nan = mask_hfac.copy()
mask_nan[mask_nan == 0] = np.nan
plot2D(is_plot, mask_nan)
annual = np.nanmean(kext[92 : 92 + 364, :, :], axis=0)
annual = np.squeeze(annual) * mask_nan
print("Shape annual", np.shape(annual))

plot2D(is_plot, annual)

# ora interpolo sulla griglia GoT
mlat = np.arange(yc_origin, yc_origin + ny_got * delY, delY)
mlon = np.arange(xc_origin, xc_origin + nx_got * delX, delX)
clat, clon = np.meshgrid(mlat, mlon)
print("Shape lat and lon", np.shape(mlat), np.shape(mlon))

fh_lon = open(f"{strdir_grid}Computational_XC", "rb")
lon = np.fromfile(fh_lon, dtype=">f4").reshape((ny, nx))
fh_lon.close()
fh_lat = open(f"{strdir_grid}Computational_YC", "rb")
lat = np.fromfile(fh_lat, dtype=">f4").reshape((ny, nx))
fh_lat.close()

lat_nan = lat * mask_nan
lon_nan = lon * mask_nan

varint = griddata(
    (lon[~np.isnan(mask_nan)], lat[~np.isnan(mask_nan)]),
    annual[~np.isnan(mask_nan)],
    (clon, clat),
    method="linear",
)
print("Shape varint", np.shape(varint))

# plot2D(is_plot,clat)
# plot2D(is_plot,clon)
# plot2D(is_plot,lat_nan)
# plot2D(is_plot,lon_nan)
plot2D(is_plot, varint.T)

if is_plot:
    plt.figure()
    plt.pcolor(annual)
    plt.colorbar()
    # Set color bar limits
    plt.clim(0.06, 0.22)
    plt.show()

    plt.figure()
    plt.pcolor((varint.T * mask_got))
    plt.colorbar()
    plt.clim(0.06, 0.22)
    plt.show()

fw = open(ptd.kext_path + "Kext_GoT_iNEST_365.dat", "wb")
for _ in range(365):
    fw.write(varint.astype(np.float32).tobytes())
fw.close()
print("Done Kext_for_GoT_iNEST_V2")

File: legacy_code/m_code/Kext_for_GoT_iNEST_V2.m
clear all
close all

dataFile='/share/data/vdibiagio/MITgcm_BFM/lightExt/xESP_forc_days1to1920.bin';

nx=570;
ny=264;

strdirGrid='/share/scratch/vdibiagio/MITgcmBFM/CLR_365days/results_merged/';
fhfacC = fopen([strdirGrid,'Computational_hFacC'], 'r','b');
MASKhFac=fread(fhfacC,[nx,ny],'float32');
MASKhFac(MASKhFac~=1)=0;
Kext=zeros(1921,nx,ny);
fKext = fopen(dataFile, 'r','b');
for idays=1:1921
    A= fread(fKext,[nx,ny],'float32');
    Kext(idays,:,:)=A.*MASKhFac;
end

MASKNan=MASKhFac;
MASKNan(MASKNan==0)=NaN;
ANNUAL=mean(Kext(93:93+364,:,:),1);
ANNUAL=squeeze(ANNUAL).*MASKNan;

close all
B=(ANNUAL);
figure;pcolor(B');shading flat;colorbar

% ora interpolo sulla griglia GoT
ycOrigin=45.4;
xcOrigin=13.22;
delX=1/768;
delY=1/768;
nxGoT=450;
nyGoT=300;

mlat=ycOrigin:delY:ycOrigin+(nyGoT-1)*delY;
mlon=xcOrigin:delX:xcOrigin+(nxGoT-1)*delX;
[CLAT,CLON]=meshgrid(mlat,mlon);

fhLON = fopen([strdirGrid,'Computational_XC'], 'r','b');
LON=fread(fhLON,[nx,ny],'float32');
fhLAT = fopen([strdirGrid,'Computational_YC'], 'r','b');
LAT=fread(fhLAT,[nx,ny],'float32');

LATnan=LAT.*MASKNan;
LONnan=LON.*MASKNan;

% "salto" molto grande in risoluzione da Med a GoT_iNEST...
varint=griddata(LON(~isnan(MASKNan)),LAT(~isnan(MASKNan)),...
    ANNUAL(~isnan(MASKNan)),CLON,CLAT,'linear'); % nearest è troppo "a blocchi"...

close all
figure;pcolor(CLAT');shading flat;colorbar;
figure;pcolor(CLON');shading flat;colorbar;
figure;pcolor(LATnan');shading flat;colorbar;
figure;pcolor(LONnan');shading flat;colorbar;
figure;pcolor(varint');shading flat;colorbar;
figure;pcolor(ANNUAL');shading flat;colorbar;
caxis([0.06 0.22]);

maskFileGoT='/share/data/squerin/esperimenti/GoT_iNEST/bathymetry/bathydata_GoT_iNEST_V2/hFacC.data';
fr2=fopen(maskFileGoT,'r','l');
maskGoT=fread(fr2,[nxGoT,nyGoT],'float32');
figure;pcolor((varint.*maskGoT)');shading flat;colorbar;
caxis([0.06 0.22])

fw=fopen('Kext_GoT_iNEST_365.dat','w','l');
for timestep=1:365
    fwrite(fw,varint,'float32');
end

fclose('all')
